### ISSUE TITLE
Disable waving leaves for trees next to the player

### ISSUE THREAD

**Issue by kneekoo** [minetest/minetest#10929]
As I have a tiny issue with motion sickness, waving leaves **up close** are bit hard to watch while cutting down trees. I have no problem whatsoever with other waving vegetation, but only the very close stuff, so currently I can enable waving leaves and make sure I look elsewhere while cutting trees, or disable the feature completely.

It would be great for the engine to either disable the waving of leaves that are within a certain radius from the player, or have an extra setting for people having the same issue - or support this internally for game makers to play with it as they see fit. This is somewhat in line with the accessibility features that I mentioned in #10666, so in the future it could be part of that.

Waving stuff is generally nice to have, making the game more appealing, so I'd love a middle ground solution for this.

**Comment by v-rob** [comment]
It seems to me that the best idea is probably to just disable waving leaves entirely if there's a problem with it; the game doesn't look much worse without it.  This sounds like a rather niche setting to have to support that the vast majority of people wouldn't use.

**Comment by kneekoo** [comment]
Even if it's not a setting in the UI, as long as the engine supports changing the waving feature, the game/mod developers can use it for more than what I just mentioned here. I can easily see a player entering in an "enchanted room/cave" and the vegetation inside starts waving, or you'd use red stone mechanics to toggle waving by flipping a switch or in the presence of daylight.

People can get creative with this, while also covering some health-related issues for those who need it.

**Comment by randomMesh** [comment]
A double ended slider would be nice.

`Waving Distance:   Min [====X==============X===] Max`

**Comment by 0-afflatus** [comment]
There is already an option to disable waving leaves.

**Comment by kneekoo** [comment]
I know, read the issue.

**Comment by lhofhansl** [comment]
@hecktest Is a fan of having all these server controlled or at least server controllable. And I would agree.

**Comment by randomMesh** [comment]
> having all these server controlled

This contradicts the MVC pattern. The controller should not tell the view how to render its model.

**Comment by Wuzzy2** [comment]
> hecktest Is a fan of having all these server controlled or at least server controllable. And I would agree.

Note that it is already server-controlled ... indirectly. If `waving = nil` for the node, it never waves.

**Comment by kneekoo** [comment]
So this means a game can control waving vegetation at block level? That would allow a game to have its own settings for the users to adjust.

**Comment by v-rob** [comment]
> as long as the engine supports changing the waving feature

This is a ticklish business since it requires the ability to change node definitions at runtime.  See #36 and all associated issues for more information on why this is tricky and not feasible at the present time.

Simply disabling waving leaves near to the player as a setting is possible, but as I said, that seems like a feature that very few people would use, so I don't know if it's worth having to support such a niche feature.

**Comment by kneekoo** [comment]
Holy rabbit hole, that's a deep one. But as I suggested, there are other game features that can be born out of this of the engine supports this.

Now I don't mean to annoy anyone, but just because the majority doesn't have issues, it doesn't mean we're supposed to forever ignore the needs of the few(er). The [USA alone has about 21 million people](https://www.cdc.gov/ncbddd/disabilityandhealth/infographic-disability-impacts-all.html) with some kind of disability, and of course the number is much higher summing it up globally.

I understand this presents some challenges at the moment, but it's still something to consider because it can and will benefit many others as well. The enchanted room/cave idea is a nice use case for this engine feature. Disabled or not, it would be cool to see a room/cave coming to life.

**Comment by sfan5** [comment]
Don't shaders have position information available? If so one could implement this feature today by just editing the GLSL files shipped with Minetest, no recompiling necessary.

**Comment by randomMesh** [comment]
> one could implement this feature today by just editing the GLSL files shipped with Minetest

This should be possible for sure - if you have experience with GLSL that i assume 90% of the Minetest users don't have.

Lets make it easy for them and just add a fricking slider to the gui.

**Comment by v-rob** [comment]
> The enchanted room/cave idea is a nice use case for this engine feature

Right, but as I said, that feature is _extremely non-trivial_ because it requires dynamic nodes.  That's a really tall order.  It cannot be done without a lot of work, and if it was done, it would affect much more than just waving things.

The _setting_, on the other hand, can be done easily enough, but I think it would be better to _just disable waving leaves if you have a problem with them_.  They are a non-essential feature.  I have nothing against disabled people; I would oppose other niche settings as well, especially when they have an easy workaround by disabling the non-essential feature.  If someone implements this, fine, I won't argue.  But I don't consider it worth spending time on.

**Comment by mk-pmb** [comment]
Could someone who knows about GLSL share what would need to be adjusted there? Then others can make a little utility program to help affected users modify the necessary files using a double slider.

**Comment by retrotails** [comment]
```diff
diff --git a/client/shaders/nodes_shader/opengl_vertex.glsl b/client/shaders/nodes_shader/opengl_vertex.glsl
index 42fae7612..494b20c98 100644
--- a/client/shaders/nodes_shader/opengl_vertex.glsl
+++ b/client/shaders/nodes_shader/opengl_vertex.glsl
@@ -152,7 +152,7 @@ float snoise(vec3 p)
 void main(void)
 {
        varTexCoord = inTexCoord0.st;
-
+       float strength = min(1.0,max(0.0,  (mWorldView * inVertexPosition).z/50.0 - 0.25  ));
        float disp_x;
        float disp_z;
 // OpenGL < 4.3 does not support continued preprocessor lines
@@ -160,10 +160,10 @@ void main(void)
        vec4 pos2 = mWorld * inVertexPosition;
        float tOffset = (pos2.x + pos2.y) * 0.001 + pos2.z * 0.002;
        disp_x = (smoothTriangleWave(animationTimer * 23.0 + tOffset) +
-               smoothTriangleWave(animationTimer * 11.0 + tOffset)) * 0.4;
+               smoothTriangleWave(animationTimer * 11.0 + tOffset)) * 0.4 * strength;
        disp_z = (smoothTriangleWave(animationTimer * 31.0 + tOffset) +
                smoothTriangleWave(animationTimer * 29.0 + tOffset) +
-               smoothTriangleWave(animationTimer * 13.0 + tOffset)) * 0.5;
+               smoothTriangleWave(animationTimer * 13.0 + tOffset)) * 0.5 * strength;
 #endif
 
        vec4 pos = inVertexPosition;
```
this smoothly fades the waving; the closer stuff is, the less it waves.  
just dropping this here for anyone that wants to enable it manually.  
for me on linux the file to modify is located in `/usr/share/minetest/client/shaders/nodes_shader/opengl_vertex.glsl`
